crates/iceberg/src/io.rs

```diff
@@ -0,0 +1,110 @@
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! This module contains io implementation.
+
+use crate::error::Result;
+use opendal::{Operator, Reader, Writer};
+
+/// FileIO implementaion.
+///
+/// This is a wrapper of opendal operator, which already provided an abstraction over all kinds of storage.
+pub struct FileIO {
+    op: Operator,
+}
+
+impl FileIO {
+    /// Creates from opendal operator.
+    pub fn new(op: Operator) -> Self {
+        Self { op }
+    }
+
+    /// Deletes file.
+    pub async fn delete(&self, path: impl AsRef<str>) -> Result<()> {
+        Ok(self.op.delete(path.as_ref()).await?)
+    }
+
+    /// Creates input file.
+    pub fn new_input(&self, path: impl Into<String>) -> Result<InputFile> {
+        Ok(InputFile {
+            op: self.op.clone(),
+            path: path.into(),
+        })
+    }
+
+    /// Creates output file.
+    pub fn new_output(&self, path: impl Into<String>) -> Result<OutputFile> {
+        Ok(OutputFile {
+            op: self.op.clone(),
+            path: path.into(),
+        })
+    }
+}
+
+/// Input file implementation.
+pub struct InputFile {
+    op: Operator,
+    path: String,
+}
+
+impl InputFile {
+    /// Fully qualified of path.
+    pub fn location(&self) -> &str {
+        &self.path
+    }
+
+    /// Check if file exists.
+    pub async fn exists(&self) -> Result<bool> {
+        Ok(self.op.is_exist(&self.path).await?)
+    }
+
+    /// Returns [`opendal::Reader`] for reading.
+    pub async fn reader(&self) -> Result<Reader> {
+        Ok(self.op.reader(&self.path).await?)
+    }
+}
+
+/// Output file implementaion.
+pub struct OutputFile {
+    op: Operator,
+    path: String,
+}
+
+impl OutputFile {
+    /// Full qualified of path.
+    pub fn location(&self) -> &str {
+        &self.path
+    }
+
+    /// Checks if file exists.
+    pub async fn exists(&self) -> Result<bool> {
+        Ok(self.op.is_exist(&self.path).await?)
+    }
+
+    /// Converts into [`InputFile`].
+    pub fn to_input_file(self) -> InputFile {
+        InputFile {
+            op: self.op,
+            path: self.path,
+        }
+    }
+
+    /// Returns [`opendal::Writer`] for writing.
+    pub async fn writer(&self) -> Result<Writer> {
```

Comment:
Should we consider to return `impl Writer` rather than opendal::Writer?🤔

Reply:
Good point!